specs/logging_implementation.adoc

```diff
@@ -0,0 +1,308 @@
+= Generic message mechanism for logs and alerts in Tendrl"
+
+Logs and alerts need to be treated as messages emitted by Tendrl.
+
+== Problem description
+* The logs, alerts and operations updates in Tendrl are handled differently
+  and there is no standard mechanism which makes sure no internal details are
+  exposed as part of alerts/notifications sent to end user.
+* Ideally logs messages, operation updates and alerts are messages which
+  are emitted from an application.
+
+== Use Cases
+
+Logs and alerts need to be treated as messages.
+
+== Proposed change
+* Logs, alerts and operations need to be treated as messages emitted by Tendrl modules.
+* A schema needs to be defined for emitted messages.
+* Message priorities, types etc. need to be defined at the time of emission.
+* The integration/bridge applications must pass the message as per schema.
+* For messages that serve as notifications to the administrator (or alerts),
+  exact error message must be hardcoded by the developers. No backtraces or
+  system generated error messages should be sent as notifications to the
+  administrator.
+* Notifications sent to the administrator should provide sufficient context to
+  enable debugging of the problem via investigation into the log files.
+* Notifications and alerts need to be delivered to the node agent alert sockets.
+* Common module needs to implement a logging library that hides all the implementation
+  details and configuration options from any of the components using the library.
+  Components should be able to blindly emit messages to the Message Handler module
+  in Tendrl and it should be able to deliver them to either syslog, log files,
+  alert sockets etc. depending upon the metadata and configuration.
+* Syslog must be the default logging configuration. Logging to individual
+  configuration files should be thoroughly discouraged.
+
+== Alternatives
+
+* None
+
+== Data model impact
+
+* Store the operation details in etcd:
+```
+1. Operation messages with inorder_id
```

Comment:
The etcd message directory locations are missing. Add them.

For job specific updates, the following structure is expected:
* `/queue/<job_id>` where `job_id` is the UUID of the job with no prefixes or suffixes.
* `/queue/<job_id>/parameters` containing all the parameters for the job.
* `/queue/<job_id>/messages` containing the sequentially generated keys, each containing a single message pertaining to the job.

In addition to these, there needs to be a specific file that updates with the job's overall status and marks the job as finished, either successfully or with errors, with the error details etc. @anivargi your inputs are needed here. The concern is to avoid multiple queries per poll for a job, if possible.